src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,338 @@
+        {
+            if (lastParameter.Type.Equals(cancellationTokenType) &&
+                lastParameter.IsOptional) // Has a default value being used
+            {
+                // Find out if the ct argument is using the default value
+                // Need to check among all arguments in case the user is passing them named and unordered (despite the ct being defined as the last parameter)
+                return AnyArgument(
+                    arguments,
+                    a => a.Parameter.Equals(cancellationTokenType) && a.ArgumentKind == ArgumentKind.DefaultValue);
+            }
+            return false;
+        }
+
+        // Checks if the method has a `params CancellationToken[]` argument in the last position and ensure no ct is being passed.
+        private static bool InvocationIsUsingParamsCancellationToken(
+            IParameterSymbol lastParameter,
+            ImmutableArray<IArgumentOperation> arguments,
+            INamedTypeSymbol cancellationTokenType)
+        {
+            if (lastParameter.IsParams &&
+                   lastParameter.Type is IArrayTypeSymbol arrayTypeSymbol &&
+                   arrayTypeSymbol.ElementType.Equals(cancellationTokenType))
+            {
+                IArgumentOperation? paramsArgument = arguments.FirstOrDefault(a => a.ArgumentKind == ArgumentKind.ParamArray);
+                if (paramsArgument?.Value is IArrayCreationOperation arrayOperation)
+                {
+                    // Do not offer a diagnostic if the user already passed a ct to the params
+                    return arrayOperation.Initializer.ElementValues.IsEmpty;
+                }
+            }
+
+            return false;
+        }
+
+        // Check if there's a method overload with the same parameters as this one, in the same order, plus a ct at the end.
+        private static bool MethodHasCancellationTokenOverload(
+            IMethodSymbol method,
+            ITypeSymbol cancellationTokenType,
+            [NotNullWhen(returnValue: true)] out IMethodSymbol? overload)
+        {
+            ImmutableArray<ISymbol> overloads = method.ContainingType.GetMembers(method.Name);
+            System.Collections.Generic.List<IMethodSymbol> ofType = overloads.OfType<IMethodSymbol>().ToList();
+            overload = ofType.FirstOrDefault(methodToCompare =>
+                HasSameParametersPlusCancellationToken(cancellationTokenType, method, methodToCompare));
+
+            return overload != null;
+
+            // Checks if the parameters of the two passed methods only differ in a ct.
+            static bool HasSameParametersPlusCancellationToken(ITypeSymbol cancellationTokenType, IMethodSymbol originalMethod, IMethodSymbol methodToCompare)
+            {
+                // Avoid comparing to itself, or when there are no parameters, or when the last parameter is not a ct
+                if (originalMethod.Equals(methodToCompare.Name) ||
```

Comment:
Do we have a unit test where you expect `originalMethod` to be same as `methodToCompare`?
```suggestion
                if (originalMethod.Equals(methodToCompare) ||
```

Reply:
Yes. When calling `method.ContainingType.GetMembers(method.Name).OfType<IMethodSymbol>().FirstOrDefault(...)`, among these members we will always find the one we are currently comparing.

Reply:
The current check `originalMethod.Equals(methodToCompare.Name)` will never succeed as you are comparing a method symbol with method name.